### ISSUE TITLE
Proposal: Security warnings/errors

### ISSUE THREAD

**Issue by maxlinc** [mitchellh/vagrant#4241]
I'm forking this off of #4236 to propose a compromise that maintains backwards compatibility but drives towards secure defaults.

TL;DR: I think vagrant should start warning about insecure settings like using plaintext winrm, unless the user sets `config.allow_insecure = true`. In the next major release, Vagrant should raise errors for insecure settings, unless `config.allow_insecure = :warn` or `config.allow_insecure = true`.

On #4326 @sneal wrote:

> This should default to :plaintext otherwise existing users will be angry at us for breaking their boxes.

I'm concerned because that is insecure. People often refer to WinRM as "SSH for Windows", but if you use the plaintext transport than it's more like telnet for Windows - insecure. This could be suitable for local vagrant providers, but is asking for trouble if the default for remote providers is insecure. That's why I wrote on the same issue:

> I think the default for no_ssl_peer_verification should be false. I think it's better to be secure by default and let users opt-out of security if they're not worried about using self-signed certificates for testing.

So there's a conflict here - on one hand it's better to opt-out of security than be forced to opt-into it. On the other hand, forcing an opt-out now would break backwards compatibility.

So here's my proposal:
1. Add a flag like: `config.allow_insecure` or `config.winrm.allow_insecure`
2. Allowable values should be `true`, `false`, or `:warn`
3. The default value for the current vagrant version should be `:warn`, and the warnings should note that in a future version of vagrant it will be `false`, so users should address security warnings or explicitly set the flag to `true` or `:warn`.
4. It's worth considering if vagrant should allow insecure connections to localhost, or if the provider indicates it does not need extra security (e.g. virtualbox probably doesn't need to be secured, but aws and rackspace should prefer secure connections)

I proposed a generic `config.allow_insecure` flag because there are security warnings for each transport, not just winrm. For example, in vagrant-rackspace we have used or proposed both of these for vagrant-rackspace:

``` yaml
    warn_insecure_ssh: |-
      Warning! By not specifying a custom public/private keypair,
      Vagrant is defaulting to use the insecure keypair that ships with
      Vagrant. While this isn't much of a big deal for local development,
      this is quite insecure for remote servers. Please specify a custom
      public/private keypair.
    warn_insecure_winrm: |-
      Warning! Vagrant is using plaintext communication for WinRM. While
      this isn't much of a big deal for local development, this is quite
      insecure for remote servers. Please configure WinRM to use SSL.
```

And it's easy to imagine similar warnings for NFS, SMB, or other ssh or winrm settings like disabling peer verification.

**Comment by chrisroberts** [comment]
Hi there,

We would really like this, but this issue has been open for over a year with no one working on it, Leaving it open is unfortunately making our issue count look higher than it is. I’m going to close this and if someone wants to work on it I’d still be open to a PR!

Cheers!
